Test Suite for the new Feature Support APIs

Sanithovski opened the pull request:
This PR adds a test suite for the new CD3DX12FeatureSupport class to ensure its correctness (located in googletest/feature_support_test.cpp). The test suite is created using Google Test (added as a submodule) and runs against a mock device class (googletest/MockDevice.hpp). 

The mock device class implements the ID3D12Device interface. All functions except `GetNodeCount` and `CheckFeatureSupport` does nothing. `CheckFeatureSupport` is referenced from the internal `CDevice` class with all internal functions and unreleased features removed. Instead, the mock device class contains local variables that can be modified to simulate different feature support and capability return values for each feature check. 

There are several different tests for each feature in the test suite:
* Basic test: Set arbitrary return values for each field in the feature support struct and check if the new API returns these values correctly.
* Unavailable test: Mimic the situation where the device doesn't recognize the feature and only returns E_INVALIDARG. The new API should return default values for the corresponding fields. Only applies to features from `D3D12Options1`.
* Multicore test: Simulate device with more than one graphics core. Applies to per-core features such as `Architecture and `ProtectedResourceSessionTypes`.
* Negative test: Test features involving errors other than E_INVALIDARG or special fail cases.
* Forwarding Test: For certain feature checks, the new API only forwards the input to the old API and returns the result to the user. The new API does not initialize these features in the Init() function or is not responsible for failures. In these cases, only test if the device receives the correct input values and the user receives the correct output values. The tests use arbitrary input and return values that doesn't reflect real device response. This applies to more complicated features including `FormatSupport`, `FormatInfo`, `MultisampleQualityLevels`, `QueryMetaCommand`, etc.
* Duplicate Test: Ensures that a capability is checked correctly when it's included in more than one feature support struct, especially if one of the features is unavailable.
* There's one system test in the end to that contains all feature checks.

I'm still writing the manual tests that runs against the real device. Once it's finished it will be added to this PR.

Sanithovski commented on googletest/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+
+project(DirectX-Headers-GoogleTest-Suite CXX)
+add_executable(Feature-Support-Test feature_support_test.cpp)
+target_link_libraries(Feature-Support-Test DirectX-Headers DirectX-Guids d3d12 dxcore gtest_main)
```
Google tests works fine with CMake, but when it targets Visual Studio I have to manually change the code generation method from `MDd` to `MTd` in order for it to compile. I checked online and it can be fixed by adding a compile flag to the CMakeFile. Should I add this flag to our project?

shawnhar replied:
Seems like a good idea.

Sanithovski replied:
Added the compiler flags in `googletest/CMakeFiles.txt`. The workaround comes from [CMake community wiki](https://gitlab.kitware.com/cmake/community/-/wikis/FAQ#how-can-i-build-my-msvc-application-with-a-static-runtime). Not the most elegant solution, but should eliminate the need to tweak the settings manually in Visual Studio.

Sanithovski commented on googletest/MockDevice.hpp:
```diff
@@ -0,0 +1,1247 @@
+        UINT FeatureSupportDataSize
+    ) override 
+    {
+        switch( Feature )
```
I removed features that are not present in this repo's `d3d12.h` and removed all internal query functions, but I want to double check I'm not leaking any code that's not prepared for release.

Sanithovski commented on googletest/feature_support_test.cpp:
```diff
@@ -0,0 +1,1313 @@
+TEST_F(FeatureSupportTest, FormatSupportPositive)
+{
+    device->m_FormatSupport1 = D3D12_FORMAT_SUPPORT1_BUFFER | D3D12_FORMAT_SUPPORT1_TEXTURE3D | D3D12_FORMAT_SUPPORT1_SHADER_SAMPLE_COMPARISON | D3D12_FORMAT_SUPPORT1_DECODER_OUTPUT;
+    device->m_FormatSupport2 = D3D12_FORMAT_SUPPORT2_UAV_ATOMIC_ADD | D3D12_FORMAT_SUPPORT2_TILED;
```
The return value for format support (as well as other functions that involves COMMAND_ID or DXGI_FORMAT) is arbitrary. The test suite only makes sure that input and output values are forwarded correctly. In the manual test suite I will use meaningful data for testing.

shawnhar replied:
This would be a good note to capture in a comment.

General principle: any time you find yourself writing comments on a pull request to explain the code, and that comment refers to how the code in question will always be rather than just some temporary state that will change soon, consider adding a source comment instead so that the PR comment is not needed.

Sanithovski replied:
I've added this as an comment in the new commit.

Sanithovski commented on googletest/feature_support_test.cpp:
```diff
@@ -0,0 +1,1313 @@
+    // TODO: Decide on expected behavior
+    EXPECT_EQ(features.ProtectedResourceSessionTypes(0).size(), 2);
+    EXPECT_EQ(features.ProtectedResourceSessionTypes()[0], GUID());
+    EXPECT_EQ(features.ProtectedResourceSessionTypes()[1], GUID());
```
I wonder what is the expected behavior for ProtectedResourceSessionTypes if the query fails. Right now it will still initialize the returned array (vector) with the size returned from ProtectedResourceSessionTypeCount and each entry would be an empty GUID object.

shawnhar replied:
Good question!  Returning an empty vector might be better in this case?

Sanithovski replied:
I agree. Returning an empty vector is now the default behavior of this API. Resolved.

Sanithovski commented on googletest/feature_support_test.cpp:
```diff
@@ -0,0 +1,1313 @@
+    device->m_TileBasedRenderer[0] = false;
+    device->m_UMA[0] = true;
+    device->m_CacheCoherentUMA[0] = false;
+    device->m_IsolatedMMU[0] = true; // Notice that if architecture1 is not available, IslatedMMU cap should not be present
```
If Architecture1 is unavailable, the API will use results from the old Architecture for the corresponding fields. `IsolatedMMU` is not included in the old Architecture struct, so it should return false in this case.

Sanithovski reviewed (commented):
Here are some details or open issues you may be interested in. :)

shawnhar commented on googletest/feature_support_test.cpp:
```diff
@@ -0,0 +1,1313 @@
+}
+
+// Test on devices with more than one graphics node
+TEST_F(FeatureSupportTest, ArchitectureMulticore)
```
Naming nitpick: the "Multicore" naming convention used for this category of tests doesn't fit with the usage of "node" terminology elsewhere in the API.   Consider naming these tests "Multinode" instead?

Sanithovski replied:
Right. I'm using multicore because it's the first word that pops in my head :). I will change it in the next commit.

Sanithovski replied:
Resolved by renaming the tests.

Sanithovski commented:
Made some updates in recent commits:
- Added verification statements to the `test/feature_check_test.cpp` to check if the new API returns the same results as the original `CheckFeatureSupportAPI`, or default values when the feature is unsupported.
- Renamed "Multicore" tests to "Multinode".
- Added inline comments for the macros and test structures
- Renamed Feature Level check API from "HighestSupportedLevel" to "MaxLevelSupported" to conform with the original field name.
- Modified test suite CMakeFiles.txt to use `MTd` instead of `MDd` for C++ code generation. This allows Google Test to compile with the rest of the code without manual configuration in Visual Studio.
